cli/cmd/logs.go

```diff
@@ -0,0 +1,208 @@
+package cmd
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	"github.com/spf13/cobra"
+	"github.com/ttacon/chalk"
+	"k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type logFilter struct {
+	targetPod           v1.Pod
+	targetContainerName string
+}
+
+type logCmdOpts struct {
+	kubeAPI          *k8s.KubernetesAPI
+	k8sClient        *http.Client
+	controlPlanePods *v1.PodList
+	clientset        *kubernetes.Clientset
+	*logFilter
+}
+
+type ColorPicker struct {
+	m               map[string]chalk.Color
+	mu              sync.Mutex
+	availableColors []chalk.Color
+	lastUsedColor   int
+}
+
+func (c *ColorPicker) pick(id string) chalk.Color {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if color, ok := c.m[id]; !ok {
+		if c.lastUsedColor > len(c.availableColors)-1 {
+			c.lastUsedColor = 1
+		}
+		newColor := c.availableColors[c.lastUsedColor]
+		c.m[id] = newColor
+		c.lastUsedColor += 1
+		return newColor
+	} else {
+		return color
+	}
+
+}
+
+func newColorPicker() *ColorPicker {
+	return &ColorPicker{
+		m: map[string]chalk.Color{},
+		availableColors: []chalk.Color{
+			chalk.Yellow,
+			chalk.Red,
+			chalk.Cyan,
+			chalk.Green,
+			chalk.Magenta,
+			chalk.White,
+		},
+	}
+}
+
+func newLogOptions(args []string, containerFilter, kubeconfigPath, kubeContext string) (*logCmdOpts, error) {
+	kubeAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := kubeAPI.NewClient()
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(kubeAPI.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	controlPlanePods, err := clientset.
+		CoreV1().
+		Pods(controlPlaneNamespace).
+		List(meta_v1.ListOptions{})
+
+	filterOpts, err := validateArgs(args, controlPlanePods, containerFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	return &logCmdOpts{
+		kubeAPI,
+		client,
+		controlPlanePods,
+		clientset,
+		filterOpts,
+	}, nil
+}
+
+func newCmdLogs() *cobra.Command {
+
+	var containerFilter string
+
+	cmd := &cobra.Command{
+		Use:   "logs (COMPONENT) [flags]",
+		Short: "Prints logs for controller components",
+		Long:  `Prints logs for controller components`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts, err := newLogOptions(args, containerFilter, kubeconfigPath, kubeContext)
+
```

Comment:
I would like to wait for PR #2033 to land since that will allow me to only check that the Kubernetes API is reachable. `validatePublicAPIClient looks like it does that and checks to see that the control plane is up, which we don't really need to check for with this particular command.